_resources/port1.0/group/npm-1.0.tcl

```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+#
+# This PortGroup is mainly used for npm ports (e.g. npm7). It helps setting up conflicts/dependencies/etc.
+#
+# Usage:
+#
+#     PortGroup            npm 1.0
+#     npm.setup            version
+#
+# It also makes publically accessible which node version each NPM/yarn depends on.
+# This can be accessed via [node_version NPM_VERSION]
+# e.g. npm7: [node_version 7] returns 16 for nodejs16
+#      yarn: [node_version yarn]
+
+
+# Set the minimum and maximum versions of npm
+# Required for livecheck and for setting conflicts
+# Also used by javascript portgroup
+options npm.maximum_version npm.minimum_version
+default npm.maximum_version 7
+default npm.minimum_version 3
+
+
+# This publically makes accessible which node version goes with which npm/yarn.
+# N.B. the npm.major of yarn is "yarn"
+proc node_version {npm.major} {
+    global os.major
+
+    # The platform is already darwin, so no need to double check
+    if {${npm.major} <= 5 || ${os.major} < 13} {
+        return 8
+    } else {
+        return 16
+    }
+}
+
+use_configure       no
+
+# Set all npm-specific attributes in npm.setup
+# This is only used by npm-based ports
+options npm.major npm.version
+proc npm.setup {npm_version} {
+    global prefix destroot distname conflicts
+    global npm.major npm.version npm.maximum_version npm.minimum_version
+    global PortInfo
+
+    npm.version         ${npm_version}
```

Comment:
`.setup` procedures aren't great. The existing portgroups where we have them should really be redesigned to eliminate them. Consider how the python-1.0 and php-1.1 portgroups work without needing setup procedures.

How about instead of calling `npm.setup`, make each npm* port set `npm.version`, and make the portgroup do its setup process when `npm.version` is set?

Reply:
In fact I'm not sure why a separate `npm.version` variable is needed. There's already the `version` variable. Why not use that?

Reply:
> .setup procedures aren't great. The existing portgroups where we have them should really be redesigned to eliminate them. Consider how the python-1.0 and php-1.1 portgroups work without needing setup procedures.

I see. The unpublished javascript port that I've been working on makes use of a `js.setup` structure, so I might have to rethink that.

> How about instead of calling npm.setup, make each npm* port set npm.version, and make the portgroup do its setup process when npm.version is set?

That's a great idea.

> In fact I'm not sure why a separate npm.version variable is needed. There's already the version variable. Why not use that?

I maybe should've documented that better. I wanted the `npm.setup` to have an attribute called `npm.version`, which sets the version number. This is similar to how `github.setup` sets the version number for you.

However, as you mentioned, there are better approaches that I should consider.